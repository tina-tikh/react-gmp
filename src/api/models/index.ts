export * from './movies';
export * from './searchParams';
