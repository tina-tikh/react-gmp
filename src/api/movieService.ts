import { Movie, Movies, SearchParams } from '../store/types';

const baseUrl = 'http://reactjs-cdp.herokuapp.com/movies';

export const getMovies = (opts: SearchParams): Promise<Movies> => {
  const url = new URL(baseUrl);

  Object.keys(opts).forEach(key => url.searchParams.append(key, (opts as any)[key]));
  console.log(url.searchParams);

  return fetch(url.toString())
    .then(response => response.json());
};

export const getMovie = (id: string): Promise<Movie> => {
  return fetch(`${baseUrl}/${id}`)
    .then(response => response.json());
};
