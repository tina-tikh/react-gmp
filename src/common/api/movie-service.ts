import movies from './movies-mock.json';

import { MoviesQueryResponse } from '../models';

export const retrieveMovies = (): Promise<MoviesQueryResponse> => {
  return Promise.resolve(movies);
};
