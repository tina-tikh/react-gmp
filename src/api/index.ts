import * as MovieService from './movieService';

export { MovieService };
