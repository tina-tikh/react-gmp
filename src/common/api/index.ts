import * as MovieService from './movie-service';

export { MovieService };
