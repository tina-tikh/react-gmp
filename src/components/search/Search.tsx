import * as React from 'react';
import { Component, ReactNode } from 'react';
import { connect } from 'react-redux';
import { RouteComponentProps } from 'react-router';

import { AppState, getSearchMovies, getSearchTotal } from '../../store';
import { thunkReceiveMovies } from '../../thunks';
import { Movie, SearchBy, SortBy } from '../../api';
import { ActionBar, ActionBarCaption, Header, Main, MainMessage } from '../layout';
import { ToggleGroup } from '../form';
import FilmList from '../FilmList';
import SearchBox from './SearchBox';
import SearchCaption from './SearchCaption';
import SearchHeader from './SearchHeader';
import SearchManager from './SearchManager';

interface SearchProps extends RouteComponentProps<any> {
  movies: Movie[],
  total: number,
  thunkReceiveMovies: typeof thunkReceiveMovies
}

class Search extends Component<SearchProps> {
  private searchManager: SearchManager<any>;

  constructor(props: SearchProps) {
    super(props);
    this.searchManager = new SearchManager(props.location);
  }

  componentDidUpdate(prevProps: Readonly<SearchProps>) {
    if (this.props.location !== prevProps.location) {
      const params = this.searchManager.getSearchParamsObj();
      this.props.thunkReceiveMovies(params);
    }
  }

  render(): ReactNode {
    const params = this.searchManager.getSearchParamsObj();
    const total = this.props.total;
    const movies = this.props.movies;

    return (
      <React.Fragment>
        <Header>
          <SearchHeader>
            <SearchCaption>Find your movie</SearchCaption>
            <SearchBox searchValue={params.search}
                       onSubmit={(e: string) => this.handleSubmit(e)}
            ></SearchBox>
            <ToggleGroup
              label="Search by"
              value={params.searchBy}
              values={Object.values(SearchBy)}
              valueLabels={Object.keys(SearchBy)}
              onChange={(e: string) => this.handleSearchByChange(e)}
            ></ToggleGroup>
          </SearchHeader>
        </Header>
        <Main>
          <ActionBar>
            <ActionBarCaption>{total} movie found</ActionBarCaption>
            <ToggleGroup
              label="Sort by"
              value={params.sortBy}
              values={Object.values(SortBy)}
              valueLabels={Object.keys(SortBy)}
              onChange={(e: string) => this.handleSortByChange(e)}
            ></ToggleGroup>
          </ActionBar>
          <FilmList films={movies}/>
          {total === 0 && <MainMessage>No films found</MainMessage>}
        </Main>
      </React.Fragment>
    );
  }

  handleSearchByChange(searchByValue: string): void {
    this.searchManager.setSearchBy(searchByValue as SearchBy);
    this.refresh();
  }

  handleSortByChange(sortByValue: string): void {
    this.searchManager.setSortBy(sortByValue as SortBy);
    this.refresh();
  }

  handleSubmit(q: string): void {
    this.searchManager.setQuery(q);
    this.refresh();
  }

  refresh(): void {
    this.props.history.push('/search?' + this.searchManager.queryParams);
  }
}

const mapStateToProps = (state: AppState) => ({
  movies: getSearchMovies(state),
  total: getSearchTotal(state)
});

export default connect(
  mapStateToProps,
  { thunkReceiveMovies }
)(Search);
