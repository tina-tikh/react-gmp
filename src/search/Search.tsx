import * as React from 'react';

import { ActionBar, ActionBarCaption, FilmList, Header, Main, MainMessage, Movie, ToggleGroup } from '../common';
import movies from '../common/api/movies-mock.json';
import SearchBox from './SearchBox';
import SearchCaption from './SearchCaption';
import SearchHeader from './SearchHeader';

type SearchProps = unknown;
type SearchState = {
  searchByValue: string;
  sortByValue: string;
  searchResult: { data: Movie[]; total: number };
};

class Search extends React.Component<SearchProps, SearchState> {
  searchByOpts: string[] = ['Title', 'Genre'];
  sortByOpts: string[] = ['Release', 'Rating'];

  constructor(props: any) {
    super(props);
    this.state = {
      searchByValue: this.searchByOpts[0],
      sortByValue: this.sortByOpts[0],
      searchResult: movies,
    };
  }

  render() {
    const res = this.state.searchResult;
    return (
      <React.Fragment>
        <Header>
          <SearchHeader>
            <SearchCaption>Find your movie</SearchCaption>
            <SearchBox
              onSubmit={(e: string) => this.handleSubmit(e)}
            ></SearchBox>
            <ToggleGroup
              label="Search by"
              value={this.state.searchByValue}
              opts={this.searchByOpts}
              onChange={(e: string) => this.handleSearchByChange(e)}
            ></ToggleGroup>
          </SearchHeader>
        </Header>
        <Main>
          <ActionBar>
            <ActionBarCaption>{res.total} movie found</ActionBarCaption>
            <ToggleGroup
              label="Sort by"
              value={this.state.sortByValue}
              opts={this.sortByOpts}
              onChange={(e: string) => this.handleSortByChange(e)}
            ></ToggleGroup>
          </ActionBar>
          <FilmList films={res.data} />
          {res.total === 0 && <MainMessage>No films found</MainMessage>}
        </Main>
      </React.Fragment>
    );
  }

  handleSearchByChange(searchByValue: string) {
    this.setState({
      searchByValue,
    });
  }

  handleSortByChange(sortByValue: string) {
    this.setState({
      sortByValue,
    });
  }

  handleSubmit(e: string) {
    this.setState({
      searchResult: {
        data: [],
        total: 0,
      },
    });
  }
}

export default Search;
