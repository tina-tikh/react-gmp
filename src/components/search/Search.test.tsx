import * as React from 'react';
import { mocked } from 'ts-jest/utils';
import { fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import { renderWithTheme } from '../../../test';
import { MovieService } from '../../api';
import Search from './Search';

jest.mock('react-router-dom');
jest.mock('../common/api/movie-service');

describe('<Search />', () => {
  // const moviesResponseMock: Movie[] = [
  //   {
  //     id: 123,
  //     title: '',
  //     tagline: '',
  //     vote_average: 0,
  //     vote_count: 9,
  //     release_date: '',
  //     poster_path: 'link',
  //     overview: '',
  //     budget: 0,
  //     revenue: 0,
  //     genres: ['Action', 'Adventure', 'Science Fiction'],
  //     runtime: null,
  //   },
  // ];

  // beforeEach(async () => {
  //   mocked(MovieService).retrieveMovies.mockResolvedValue(moviesResponseMock);
  // });
  //
  // it('should render search', () => {
  //   const { asFragment } = renderWithTheme(<Search />);
  //   expect(asFragment()).toMatchSnapshot();
  // });
  //
  // it('should handle search query',  () => {
  //   const query = '123';
  //   const { getByText, getByPlaceholderText } = renderWithTheme(<Search />);
  //
  //   mocked(MovieService).searchBy.mockResolvedValue([]);
  //
  //   fireEvent.change(getByPlaceholderText('Search'), { target: { value: query } });
  //   fireEvent.click(getByText('Search'));
  //
  //   expect(getByText('No films found')).toBeInTheDocument();
  // });
  //
  // it('should handle search filter', () => {
  //   const { asFragment, getByText} = renderWithTheme(<Search />);
  //   fireEvent.click(getByText('Genre'));
  //   expect(asFragment()).toMatchSnapshot();
  // });
  //
  // it('should handle sort filter', () => {
  //   const { asFragment, getByText} = renderWithTheme(<Search />);
  //   fireEvent.click(getByText('Rating'));
  //   expect(asFragment()).toMatchSnapshot();
  // });
});
