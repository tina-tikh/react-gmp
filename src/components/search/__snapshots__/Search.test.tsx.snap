// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search /> should handle search filter 1`] = `
<DocumentFragment>
  <header
    class="css-t3r2z7"
  >
    <div
      class="css-1r4e12l"
    >
      <h1
        class="css-sokrpn"
      >
        Find your movie
      </h1>
      <div
        class="css-11yzz3q"
      >
        <input
          class="css-1wf26wm"
          placeholder="Search"
          value=""
        />
        <button
          class="css-u83ny6"
        >
          Search
        </button>
      </div>
      <div
        class="css-17wzkxm"
      >
        <label
          class="css-1acepui"
        >
          Search by
        </label>
        <button
          class=" css-1f5llbs"
        >
          Title
        </button>
        <button
          class="active css-1f5llbs"
        >
          Genre
        </button>
      </div>
    </div>
  </header>
  <main
    class="css-1wufzma"
  >
    <section
      class="css-5719bz"
    >
      <span
        class="css-7j8j8h"
      >
        0 movie found
      </span>
      <div
        class="css-17wzkxm"
      >
        <label
          class="css-1acepui"
        >
          Sort by
        </label>
        <button
          class="active css-1f5llbs"
        >
          Release
        </button>
        <button
          class=" css-1f5llbs"
        >
          Rating
        </button>
      </div>
    </section>
    <ul
      class="css-17vq9x5"
    />
    <h1
      class="css-1dglaoy"
    >
      No films found
    </h1>
  </main>
</DocumentFragment>
`;

exports[`<Search /> should handle sort filter 1`] = `
<DocumentFragment>
  <header
    class="css-t3r2z7"
  >
    <div
      class="css-1r4e12l"
    >
      <h1
        class="css-sokrpn"
      >
        Find your movie
      </h1>
      <div
        class="css-11yzz3q"
      >
        <input
          class="css-1wf26wm"
          placeholder="Search"
          value=""
        />
        <button
          class="css-u83ny6"
        >
          Search
        </button>
      </div>
      <div
        class="css-17wzkxm"
      >
        <label
          class="css-1acepui"
        >
          Search by
        </label>
        <button
          class="active css-1f5llbs"
        >
          Title
        </button>
        <button
          class=" css-1f5llbs"
        >
          Genre
        </button>
      </div>
    </div>
  </header>
  <main
    class="css-1wufzma"
  >
    <section
      class="css-5719bz"
    >
      <span
        class="css-7j8j8h"
      >
        0 movie found
      </span>
      <div
        class="css-17wzkxm"
      >
        <label
          class="css-1acepui"
        >
          Sort by
        </label>
        <button
          class=" css-1f5llbs"
        >
          Release
        </button>
        <button
          class="active css-1f5llbs"
        >
          Rating
        </button>
      </div>
    </section>
    <ul
      class="css-17vq9x5"
    />
    <h1
      class="css-1dglaoy"
    >
      No films found
    </h1>
  </main>
</DocumentFragment>
`;

exports[`<Search /> should render search 1`] = `
<DocumentFragment>
  <header
    class="css-t3r2z7"
  >
    <div
      class="css-1r4e12l"
    >
      <h1
        class="css-sokrpn"
      >
        Find your movie
      </h1>
      <div
        class="css-11yzz3q"
      >
        <input
          class="css-1wf26wm"
          placeholder="Search"
          value=""
        />
        <button
          class="css-u83ny6"
        >
          Search
        </button>
      </div>
      <div
        class="css-17wzkxm"
      >
        <label
          class="css-1acepui"
        >
          Search by
        </label>
        <button
          class="active css-1f5llbs"
        >
          Title
        </button>
        <button
          class=" css-1f5llbs"
        >
          Genre
        </button>
      </div>
    </div>
  </header>
  <main
    class="css-1wufzma"
  >
    <section
      class="css-5719bz"
    >
      <span
        class="css-7j8j8h"
      >
        0 movie found
      </span>
      <div
        class="css-17wzkxm"
      >
        <label
          class="css-1acepui"
        >
          Sort by
        </label>
        <button
          class="active css-1f5llbs"
        >
          Release
        </button>
        <button
          class=" css-1f5llbs"
        >
          Rating
        </button>
      </div>
    </section>
    <ul
      class="css-17vq9x5"
    />
    <h1
      class="css-1dglaoy"
    >
      No films found
    </h1>
  </main>
</DocumentFragment>
`;
